sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -564,6 +565,16 @@ object TypeCoercion {
         NaNvl(Cast(l, DoubleType), r)
       case NaNvl(l, r) if r.dataType == NullType => NaNvl(l, Cast(r, l.dataType))
     }
+    // This group needs to be transformed in a post order
+    .transformExpressionsUp {
+      // When all inputs in [[Concat]] are binary, coerces an output type to binary
+      case c @ Concat(children, _)
```

Comment:
Create a new `TypeCoercion ` rule for it?

Reply:
yea, ok.